sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -2555,6 +2555,24 @@ class DataFrameSuite extends QueryTest
     val df = Seq(0.0 -> -0.0).toDF("pos", "neg")
     checkAnswer(df.select($"pos" > $"neg"), Row(false))
   }
+
+  test("SPARK-32851: should fail if errors happen when generating expr code") {
+    val errMsg = intercept[SparkException] {
+      withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "false") {
+        val df = spark.range(1)
+        df.select(Column(FailedCodegenExpr(df("id").expr)) :: Nil: _*).show()
+      }
+    }.getCause.getMessage
+    assert(errMsg.contains("failed to compile:"))
+  }
 }
 
 case class GroupByKey(a: Int, b: Int)
+
+case class FailedCodegenExpr(child: Expression) extends UnaryExpression with CodegenFallback {
+  private lazy val proj = FailedCodegenProjection.createObject(child :: Nil)
+  override def dataType: DataType = LongType
+  override def eval(input: InternalRow): Any = {
+    proj(input).getLong(0)
```

Comment:
This test proves that if `eval` fails, the query fails. I think we should override `doCodeGen` and produce invalid code there?

Reply:
This test proves that, if the codegen of `proj` fails, the query fails.  If `CODEGEN_FACTORY_MODE`=fallback, `proj` falls back into an interpreted one.

Reply:
Then shall we just add a test in `CodeGeneratorWithInterpretedFallbackSuite`,  to make sure that `FailedCodegenProjection` fails with default codegen mode?

Reply:
My point is that, the test here is not really an end-to-end test. It's just used to trigger `FailedCodegenProjection`

Reply:
>  It's just used to trigger FailedCodegenProjection

Yea, yes. This test just triggers the the codegen failure of `FailedCodegenProjection` in executor sides and it checks the failure correctly propagates to a driver side in `SharedSparkSession` by setting the config `CODEGEN_FACTORY_MODE`=`codegen_only` in `SparkConf`:
https://github.com/apache/spark/pull/29721/files#diff-ef6933b96996929c402e3a36296b2590R71

`CodeGeneratorWithInterpretedFallbackSuite` is placed in the catalyst package, so it is not related to adding the config in `SharedSparkSession`. But, if the existing tests in  `CodeGeneratorWithInterpretedFallbackSuite` look enough, I think it is okay to drop the added tests and just add the config `CODEGEN_FACTORY_MODE` in `SharedSparkSession`.

Reply:
Yea I think it's OK to just change the config in test.

Reply:
okay ~